### ISSUE TITLE
jquery优化

### ISSUE THREAD

**Issue by jennyWan278** [jennyWan278/issue#37]
[参考链接](http://www.cnblogs.com/xiaohuochai/p/8270561.html)
### jQuery代码优化的9种方法
* 用对选择器
    * 最快的选择器：id选择器和元素标签选择器
    * 较慢的选择器：class选择器
    * 最慢的选择器：伪类选择器和属性选择器
* 理解父子关系,从父元素中选择子元素
    * $('.child', $parent)  
     jQuery会自动把这条语句转成$.parent.find('child')，这会导致一定的性能损失。它比最快的形式慢了5%-10%
    * $parent.find('.child')  
    这条是最快的语句。.find()方法会调用浏览器的原生方法（getElementById，getElementByName，getElementByTagName等等），所以速度较快
    * $parent.children('.child') 
     这条语句在jQuery内部，会使用$.sibling()和javascript的nextSibling()方法，一个个遍历节点。它比最快的形式大约慢50%
    * $('#parent > .child')  
    这条语句是先选.child，然后再一个个过滤出父元素#parent，这导致它比最快的形式大约慢70%
    * $('#parent .child')  
    这条语句与上一条是同样的情况。但是，上一条只选择直接的子元素，这一条可以选择多级子元素，所以它的速度更慢，大概比最快的形式慢了77%
    * $('.child', $('#parent')) 
    jQuery内部会将这条语句转成$('#parent').find('.child')，比最快的形式慢了23%
* 不过度使用jQuery
    * jQuery速度再快，也无法与原生的javascript方法相比。所以有原生方法可以使用的场合，尽量避免使用jQuery。
    * 以最简单的选择器为例，document.getElementById("foo")要比$("#foo")快10多倍。
* 做好缓存
    * 使用链式写法 
$('div').find('h3').eq(2).html('Hello');
    * 选中某一个网页元素，是开销很大的步骤。所以，使用选择器的次数应该越少越好，并且尽可能缓存选中的结果，便于以后反复使用。
```javascript
//较差的写法
　jQuery('#top').find('p.classA');
　jQuery('#top').find('p.classB');
//较好的写法
   var cached = jQuery('#top');
   cached.find('p.classA');
　cached.find('p.classB');
```
* 事件委托
 javascript的事件模型，采用"冒泡"模式，也就是说，子元素的事件会逐级向上"冒泡"，成为父元素的事件。
   利用这一点，可以大大简化事件的绑定。比如，有一个表格（table元素），里面有100个格子（td元素），现在要求在每个格子上面绑定一个点击事件（click），请问是否需要将下面的命令执行100次？
    回答是不需要，我们只要把这个事件绑定在table元素上面就可以了，因为td元素发生点击事件之后，这个事件会"冒泡"到父元素table上面，从而被监听到
    这个事件只需要在父元素绑定1次即可，而不需要在子元素上绑定100次，从而大大提高性能。这就叫事件的"委托处理"，也就是子元素"委托"父元素处理这个事件
```javascript
//普通写法
$("td").on("click", function(){
　　　　$(this).toggleClass("click");
　　});

//事件委托写法
　$("table").on("click", "td", function(){
　　　　$(this).toggleClass("click");
　　});

//更好的写法，则是把事件绑定在document对象上面
　　$(document).on("click", "td", function(){
　　　　$(this).toggleClass("click");
　　});

//如果要取消事件的绑定，就使用off()方法
       $(document).off("click", "td");
```
* 少改动DOM
    * 改动DOM结构开销很大，因此不要频繁使用.append()、.insertBefore()和.insetAfter()这样的方法
    如果要插入多个元素，就先把它们合并，然后再一次性插入。根据测试，合并插入比不合并插入，快了将近10倍
    * 如果要对一个DOM元素进行大量处理，应该先用.detach()方法，把这个元素从DOM中取出来，处理完毕以后，再重新插回文档。根据测试，使用.detach()方法比不使用时，快了60%
    * 插入html代码的时候，浏览器原生的innterHTML()方法比jQuery对象的html()更快
    * 如果要在DOM元素上储存数据，写成这样
```javascript
    var elem = $('#elem');
　$.data(elem[0],key,value);
```
* 尽量少生成jQuery对象
    * 每当使用一次选择器（比如$('#id')），就会生成一个jQuery对象。jQuery对象是一个很庞大的对象，带有很多属性和方法，会占用不少资源。所以，尽量少生成jQuery对象.
* 选择作用域链最短的方法
    * Javascript的变量采用链式作用域。读取变量的时候，先在当前作用域寻找该变量，如果找不到，就前往上一层的作用域寻找该变量。这样的设计，使得读取局部变量比读取全局变量快得多
* 使用Pub/Sub模式管理事件
